compiler/src/dmd/expressionsem.d

```diff
@@ -12796,6 +12796,59 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             return;
         }
 
+        // Inline the expression, if possible.
+        BinExp be = cast(BinExp)e;
+        if (be.e1.type.isScalar() && be.e2.isIntegerExp())
+        {
+            Expression one;
+            if (be.e1.type.isIntegral()) {
+                one = new IntegerExp(e.loc, 1, be.e1.type);
+            } else {
+                one = new RealExp(e.loc, CTFloat.one, be.e1.type);
+            }
+
+            const expo = cast(sinteger_t)be.e2.toInteger();
+            // Replace e1 ^^ -1 with 1 / e1
+            if (expo == -1)
+            {
+                Expression ex = new DivExp(exp.loc, one, be.e1);
+                ex = ex.optimize(WANTvalue);
+                ex = ex.expressionSemantic(sc);
+                result = ex;
+                return;
+            }
+            // Replace e1 ^^ 0 with 1
+            else if (expo == 0)
+            {
+                Expression ex = one;
+                ex.loc = exp.loc;
+                ex = ex.expressionSemantic(sc);
+                result = ex;
+                return;
+            }
+            // Replace e1 ^^ 1 with e1
+            else if (expo == 1)
+            {
+                Expression ex = be.e1;
+                ex.loc = exp.loc;
+                ex = ex.expressionSemantic(sc);
+                result = ex;
+                return;
+            }
+            // Replace e1 ^^ 2 with e1 * e1
+            else if (expo == 2)
+            {
+                auto v = copyToTemp(STC.const_, "__powtmp", be.e1);
```

Comment:
Good that you thought about side-effects here :+1:

Reply:
Thanks!